src/main/java/com/boot/cart/service/CartService.java

```diff
@@ -16,236 +22,256 @@
+
     CartRepository cartRepository;
 
-    @Autowired
     ProductServiceClient productServiceClient;
 
-    @Autowired
     UserServiceClient userServiceClient;
 
     public CartDTO addProductToCart(String email, String productName, int quantity)
             throws InvalidInputDataException, EntityNotFoundException {
         log.info("addProductToCart - process started");
-        Product product;
+        ProductDTO productDTO;
         try {
-            product = ProductMapper.DtoToProductEntity(productServiceClient.callGetProductByProductName(productName));
+            productDTO = productServiceClient.callGetProductByProductName(productName);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Product: " + productName + " not found in the Database!");
         }
 
-        if (product.getProductStock() == 0) {
+        if (productDTO.getStock() == 0) {
             throw new InvalidInputDataException("We are sorry, but currently: " + productName + " is out of order!");
         }
 
-        if (product.getProductStock() < quantity) {
-            throw new InvalidInputDataException("You can not add more than: " + product.getProductStock() + " "
+        if (productDTO.getStock() < quantity) {
+            throw new InvalidInputDataException("You can not add more than: " + productDTO.getStock() + " "
                     + productName + " Products to your shopping cart!");
         }
-        User user;
+        UserDTO user;
         try {
-            user = UserMapper.DtoToUserEntity(userServiceClient.callGetUserByEmail(email));
+            user = userServiceClient.callGetUserByEmail(email);
         } catch (HttpClientErrorException.NotFound e) {
             throw new EntityNotFoundException("Email: " + email + " not found in the Database!");
         }
-        Cart cart = cartRepository.findByUser(user);
+        Cart cart = cartRepository.findByUserId(user.getId());
 
-        List<Product> productList;
+        List<CartEntry> cartEntries;
         double productTotal = 0;
+
         if (cart == null) {
             cart = new Cart();
-            productList = new ArrayList<>();
+            cartEntries = new ArrayList<>();
+            CartEntry cartEntry = new CartEntry();
+            cartEntry.setProductName(productName);
+            cartEntry.setQuantity(quantity);
+            cartEntry.setCart(cart);
+
+            cartEntries.add(cartEntry);
         } else {
-            productList = cart.getProductList();
+            cartEntries = cart.getEntries();
+            for (CartEntry cartEntry : cartEntries) {
+                if (cartEntry.getProductName().equals(productName)) {
+                    cartEntry.setQuantity(cartEntry.getQuantity() + quantity);
+                    cartEntry.setCart(cart);
+                    cartEntryRepository.save(cartEntry);
+                    break;
+                }
+            }
+            List<String> productNames =
```

Comment:
you are doing this logic even if you have already found the product in the cart entries. it will work but isn't optimal.
what I would do is:`Optional<CartEntry> cartEntry= cartEntries.stream().filter(entry->entry.getProductName().equals(productName)).findFirst();
if(cartEntry.isPresent()){
            //add product in cart logic
        }else{
//add product not in cart logic
}`

Reply:
Done!